tensorflow/compiler/xla/service/gpu/gpu_compiler.cc

```diff
@@ -409,6 +409,16 @@ Status GpuCompiler::OptimizeHloPostLayoutAssignment(
     pipeline.AddPass<HloPassFix<GpuTreeReductionRewriter>>();
   }
 
+  // GemmRewriter assumes that all transposes are folded into gemms, but,
+  // since commit 7d529df, this is not always true at this point.
+  // Therefore, rerun transpose folding.
+  pipeline.AddPass<TransposeFolding>(
```

Comment:
This is not rocm-specific, right? Could you have a separate PR for this, ideally with a test case demonstrating the issue? Thanks!!

Reply:
The unit tests that are affected by this issue are
```
//tensorflow/compiler/tests:matrix_inverse_op_test_gpu
//tensorflow/compiler/tests:matrix_solve_op_test_gpu
//tensorflow/compiler/tests:qr_op_test_gpu
//tensorflow/compiler/xla/client/lib:qr_test_gpu
```
I will let @ekuznetsov139 chime-in on whether or not this is ROCm-specific (it does not appear to be so, at first glance, but I do not know the details in this case)

The two commits in this PR are what is needed to make the CI job we (rocm) have for XLA, pass for the `master` branch, and hence the grouping. Can we keep it that way?

Reply:
I think this is ROCm-specific because the function (GpuCompiler::OptimizeHloPostLayoutAssignment()) is only called for ROCm. For CUDA, TF uses its override, NVPTXCompiler::OptimizeHloPostLayoutAssignment().

Reply:
I think it affects ROCm and older CUDA versions (anything before Volta.) It might affect newest CUDA as well for some tensor dimensions.

The issue is that you can't have a transpose feeding directly into a dot. See https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc#L54 

But, on newest CUDA, there's a call to CublasGemmPadForTensorCores just before the call to GemmRewriter, which inserts a padding op between the transpose and the dot.

Reply:
@cheshire gentle ping